internal/identity/manager/schedulers.go

```diff
@@ -0,0 +1,161 @@
+package manager
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/pomerium/pomerium/pkg/grpc/session"
+)
+
+type updateUserInfoScheduler struct {
+	baseCtx                context.Context
+	updateUserInfoInterval time.Duration
+	updateUserInfo         func(ctx context.Context, userID string)
+	userID                 string
+
+	reset  chan struct{}
+	cancel context.CancelFunc
+}
+
+func newUpdateUserInfoScheduler(
+	ctx context.Context,
+	updateUserInfoInterval time.Duration,
+	updateUserInfo func(ctx context.Context, userID string),
+	userID string,
+) *updateUserInfoScheduler {
+	uuis := &updateUserInfoScheduler{
+		baseCtx:                ctx,
+		updateUserInfoInterval: updateUserInfoInterval,
+		updateUserInfo:         updateUserInfo,
+		userID:                 userID,
+		reset:                  make(chan struct{}, 1),
+	}
+	ctx, uuis.cancel = context.WithCancel(context.WithoutCancel(uuis.baseCtx))
+	go uuis.run(ctx)
+	return uuis
+}
+
+func (uuis *updateUserInfoScheduler) Reset() {
+	select {
+	case uuis.reset <- struct{}{}:
+	default:
+	}
+}
+
+func (uuis *updateUserInfoScheduler) Stop() {
+	uuis.cancel()
+}
+
+func (uuis *updateUserInfoScheduler) run(ctx context.Context) {
+	ticker := time.NewTicker(uuis.updateUserInfoInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-uuis.reset:
+			ticker.Reset(uuis.updateUserInfoInterval)
```

Comment:
I think I'm still missing something— when do we need to reset the ticker?

Reply:
We reset after the user is updated. This fixes an unnecessary update in the original scheduler. When a session is refreshed it also updates the user data, in which case it would be better to reset the timer and wait another 10 minutes before doing it again. The original scheduler would keep to its 10 minute interval regardless.

Reply:
Thanks, got it.